🛠️ Refactor Sentieon Joint Call for Beta Prod

migbro opened the pull request:
I've taken the work based off of Haodong and myself from his repo, forked it, and am shaping it into a KF DRC product. I've added documentation on how to use the "easy" workflow and the "advanced," which requires a notebook.

"Easy" example run: https://cavatica.sbgenomics.com/u/d3b-bixu/kf-sentieon-join-cohort-genotyping-dev/tasks/8fabf671-b601-4a9a-8bae-a0085614cf11/

Please let me know what details I am missing

migbro committed:
:construction: refactored main wf, started reademe

migbro committed:
:pencil: added and updated readmes
:tada: added run notebook

migbro committed:
:pencil: updated docs

migbro committed:
:pencil: add mention of Sentieon software/docs

dmiller15 commented on workflow/kf-joint-cohort-call-by-chr-wf.cwl:
```diff
@@ -0,0 +1,178 @@
+        source: split_vcf_by_chr/split_vcfs
+        valueFrom: |
+          $(self.map(function(e) { return e[inputs.scatter_index] }))
+      interval:
+        source: make_output_name/out_intvl_list
+        valueFrom: $(self[inputs.scatter_index])
+      dbSNP: dbSNP
+      call_conf: call_conf
+      emit_conf: emit_conf
+      genotype_model: genotype_model
+      output_file_name:
+        source: make_output_name/out_name_list
+        valueFrom: $(self[inputs.scatter_index])
+    scatter: [scatter_index]
```
I think you can just put `interval` and `output_file_name` into the scatter here and avoid having to do the `valueFrom` statements.

dmiller15 commented on workflow/kf-joint-cohort-call-by-chr-wf.cwl:
```diff
@@ -0,0 +1,178 @@
+      threads: bcftools_cpu
+    scatter: [input_vcf]
+    out: [split_vcfs]
+  make_output_name:
+    run:
+      class: ExpressionTool
+      cwlVersion: v1.2
+      requirements:
+      - class: InlineJavascriptRequirement
+      expression: |
+        ${
+          var out_intvl_list = inputs.chr_list.contents.trim().split("\n");
+          var out_name_list = out_intvl_list.map(function (i){
+            return inputs.output_file_prefix + "_" + i.replace(/,/g, "-") + ".vcf.gz";
+          })
+          return {"out_name_list": out_name_list, "out_intvl_list": out_intvl_list};
+        }
+      inputs:
+        output_file_prefix: string
+        chr_list: {type: File, loadContents: true}
+      outputs:
+        out_name_list: 'string[]'
+        out_intvl_list: 'string[]'
+    in:
+      output_file_prefix: output_file_prefix
+      chr_list: subset_fai/chr_list
+    out: [out_name_list, out_intvl_list]
```
I think you should just put this logic into `fai_subset.cwl`

migbro committed:
:hammer: refactor steps to be a bit cleaner for PR

migbro commented:
@dmiller15 refactored a bit as request, new task example here: https://cavatica.sbgenomics.com/u/d3b-bixu/kf-sentieon-join-cohort-genotyping-dev/tasks/cdd2fc6d-8665-4764-a565-ba362a0bdaaf/#

migbro committed:
:pencil: fix all incorrect spellings of "Sentieon"

dmiller15 commented on README.md:
```diff
@@ -1,23 +1,32 @@
-An efficient, fast, and cost effective workflow for joint calling cohorts up to ~4000 individuals on the CAVATICA platform
+An efficient, fast, and cost effective workflow for joint calling cohorts up to ~4000 individuals on the CAVATICA platform.
+You will likely need to have installed [sbpack](https://pypi.org/project/sbpack/) to push the necessary apps into your project, and it is recommended to have the python packages listed in `python_pip_requirements.txt` installed.
+Both workflows have limited implementations of the [`GVGFtyper` algo](https://support.sentieon.com/manual/usages/general/#gvcftyper-algorithm) which is part of the [sentieon driver library](https://support.sentieon.com/manual/usages/general/#driver-binary)
```
```suggestion
Both workflows have limited implementations of the [`GVGFtyper` algo](https://support.sentieon.com/manual/usages/general/#gvcftyper-algorithm) which is part of the [Sentieon driver library](https://support.sentieon.com/manual/usages/general/#driver-binary).
```

dmiller15 commented on README.md:
```diff
@@ -1,23 +1,32 @@
+Both workflows have limited implementations of the [`GVGFtyper` algo](https://support.sentieon.com/manual/usages/general/#gvcftyper-algorithm) which is part of the [sentieon driver library](https://support.sentieon.com/manual/usages/general/#driver-binary)
 
-### Badges
+![data service logo](https://github.com/d3b-center/d3b-research-workflows/raw/master/doc/kfdrc-logo-sm.png)
```
Center align this as well?

dmiller15 commented on README.md:
```diff
@@ -1,23 +1,32 @@
-`<p>` tag next to the LICENSE.
+## [Easy Cohort Calling Workflow](docs/KF_COHORT_JG_SINGLE_WF.md)
+This workflow runs on the CAVATICA platform with few parameters necessary to run.
+The platform has an internal workflow scatter limit of ~2200 files, which means, **if your cohort size is greater than this, it will not run!**.
```
```suggestion
The platform has an internal workflow scatter limit of ~2200 files, which means, **if your cohort size is greater than this, it will not run!**
```

dmiller15 commented on README.md:
```diff
@@ -1,23 +1,32 @@
-### Repo Description
+## [Advanced Cohort Calling Workflow](docs/KF_NOTEBOOK_JG_WORKFLOW.md)
+This notebook can be run locally or copied to a CAVATICA Data Studio Analysis session.
+If your cohort is >2200 samples, but projected to use less than 4TB per split job and chromosome call job, you can use this.
```
> projected to use less than 4TB per split job and chromosome call job

How can I make this calculation?

dmiller15 commented on README.md:
```diff
@@ -1,23 +1,32 @@
 
-Add additional badges for CI, docs, and other integrations as needed within the
-`<p>` tag next to the LICENSE.
+## [Easy Cohort Calling Workflow](docs/KF_COHORT_JG_SINGLE_WF.md)
```
Not sure if I like the "Easy" vs "Advanced" descriptions. Like if your input is smaller than 2200 files, you should only over use the "Easy" pipeline. There's no benefit to using the "Advanced" pipeline.

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+Then, all of the gVCF chromosome fragments are split, one instance per chromosome, to a Sentieon GVCFtyper job.
+If the user supplies a dbSNP vcf at run time, the outputs will also be annotated with `rs_` IDs in the `ID` column of the VCF.
+
+![data service logo](https://github.com/d3b-center/d3b-research-workflows/raw/master/doc/kfdrc-logo-sm.png)
```
Center align?

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+# [Kids First-Sentieon Joint Cohort Calling Workflow Beta](../workflow/kf-joint-cohort-call-by-chr-wf.cwl)
+This workflow is takes a set of gVCF files from a desired cohort and an indexed FASTA reference to create a joint-called cohort.
```
```suggestion
This workflow is takes a set of gVCFs and an indexed reference FASTA to create a joint-called cohort VCF.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+# [Kids First-Sentieon Joint Cohort Calling Workflow Beta](../workflow/kf-joint-cohort-call-by-chr-wf.cwl)
+This workflow is takes a set of gVCF files from a desired cohort and an indexed FASTA reference to create a joint-called cohort.
+It will split the input gVCF files, using up to AWS 60 c5.12xlarge instances, and stack 12 gVCF files per instance to split them by chromosome.
```
```suggestion
The input gVCFs are sharded by chromosome. To accelerate the process, the sharding is done across 60 instances, each of which is processing 12 files at a time.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+# [Kids First-Sentieon Joint Cohort Calling Workflow Beta](../workflow/kf-joint-cohort-call-by-chr-wf.cwl)
+This workflow is takes a set of gVCF files from a desired cohort and an indexed FASTA reference to create a joint-called cohort.
+It will split the input gVCF files, using up to AWS 60 c5.12xlarge instances, and stack 12 gVCF files per instance to split them by chromosome.
+Then, all of the gVCF chromosome fragments are split, one instance per chromosome, to a Sentieon GVCFtyper job.
```
```suggestion
Next, gVCF shards that share a common chromosome are gathered and passed to Sentieon GVCFtyper, producing one joint VCF per chromosome.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+ - `output_file_prefix`: Since outputs will be joint calls split by chromosome, a user-defined convenience to prefix each file output with a string of their choice
+   Default is "joint_call", so output files would be named "joint_call_chr1.vcf.gz", "joint_call_chr2.vcf.gz", etc.
+## Outputs
+ - `joint_called_by_chr_vcf`: Array of joint-called files split by chromosome
```
Shouldn't we merge these files in the end?

migbro replied:
Eh, nah, most people seem to like to operate by chromosome I think. The merge step is a slow, single-threaded step that we can consider making optional...I took it out to just have a "this is what I ended up doing" workflow

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+Two main limitations currently (as of May 2024) exist for running this cohort workflow on CAVATICA:
+ - Cohort Size
+ - File input size
+Cohort size is simple - **it must be less than ~2200 samples** to run, otherwise the platform will will quit immediately with error:
```
These are getting smashed together
```suggestion
 - File input size

Cohort size **must be less than ~2200 samples** to run, otherwise the platform will will quit immediately with error:
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
````diff
@@ -0,0 +1,45 @@
+The scheduled job root is too large to be executed. Please contact support@sbgenomics.com for further assistance if needed.
+```
+AWS instances by default start with 1TB of EBS storage.
+This can be increased to 4TB; if you still run out of space, consider a smaller cohort if possible, limiting sites, or a different platform (sorry!)
````
Not sorry...
```suggestion
CAVATICA instances, by default, start with 1TB of EBS storage. The storage can be increased through the `Execution Settings` all the way up to 4TB.
If 4TB is insufficient to process your cohort, contact CAVATICA support for assistance. Alternatively, try refining the cohort or limiting the sites of interest.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+## Inputs
+### Required
+ - `reference`: Indexed FASTA file reference. Should be the same one used to create the input gVCFs
+ - `input_vcf`: Indexed gVCF files to perform cohort calling on. It's recommended that the [GATK Haplotyper caller workflow](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/kfdrc-gatk-haplotypecaller-workflow) or [Sentieon equivalent](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/kfdrc_sentieon_gvcf_wf) had been run to generate each of these inputs. If gV ...
```
For people...like me...that don't keep up with what every initialism means
```suggestion
 - `input_vcf`: Indexed gVCF files to perform cohort calling on. It's recommended that the [GATK Haplotyper caller workflow](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/kfdrc-gatk-haplotypecaller-workflow) or [Sentieon equivalent](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/kfdrc_sentieon_gvcf_wf) has been used to generate each of these inputs. If gVCFs were obtained from Gabriella Miller Kids First (GMKF), it will meet this recommendation.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+### Required
+ - `reference`: Indexed FASTA file reference. Should be the same one used to create the input gVCFs
+ - `input_vcf`: Indexed gVCF files to perform cohort calling on. It's recommended that the [GATK Haplotyper caller workflow](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/kfdrc-gatk-haplotypecaller-workflow) or [Sentieon equivalent](https://cavatica.sbgenomics.com/public/apps/cavatica/apps-publisher/kfdrc_sentieon_gvcf_wf) had been run to generate each of these inputs. If gV ...
+ - `sentieon_license`: Sentieon license server host and port in format 0.0.0.0:0000. Is set by default by the workflow, but can be changed if circumstances require it.
```
```suggestion
 - `sentieon_license`: Sentieon license server host and port in format `0.0.0.0:0000`. Is set by default by the workflow, but can be changed if circumstances require it.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+ - `dbSNP`: Indexed dbSNP file to add common variant annotation if `ID` field
+
+### Optional
+ - `fai_subset`: Parameter used to set which chromosome are to be used based ion the fasta index file. It's generally recommended to use only chr1-22,X,Y (some would say canonical), therefore this parameter is set to `24` by default, as the typical fasta index leads with the canonical chromosomes first.
```
```suggestion
 - `fai_subset`: Parameter used to set which chromosome are to be used based on the FASTA index file. It's generally recommended to use the canonical chromosomes (chr1-22,X,Y). By default, this parameter is set to `24` to capture those first 24 chromosomes of the FASTA index.
```

dmiller15 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,45 @@
+## Run tips
+ - The default 1TB storage per instance might be enough for up to a 1500 sample cohort size. To be safe, set this to at least 2TB if not more for larger cohorts in the task, documentation on this here: https://docs.sevenbridges.com/docs/set-execution-hints-at-task-level. An example would be to use the following:
+   - Instance type: `c5.12xlarge`
+   - EBS storage: `2048` m up to `4096`
```
```suggestion
   - EBS storage: `2048` up to `4096`
```

dmiller15 commented on docs/KF_NOTEBOOK_JG_WORKFLOW.md:
```diff
@@ -0,0 +1,5 @@
+# Kids First-Sentieon Joint Cohort Call Notebook
+Due to platform limitations, running the easier workflow may not be feasible when cohorts exceed 200 samples, due to scatter job size limitations.
```
200 samples? Not 2000?

dmiller15 commented on docs/KF_NOTEBOOK_JG_WORKFLOW.md:
```diff
@@ -0,0 +1,5 @@
+# Kids First-Sentieon Joint Cohort Call Notebook
+Due to platform limitations, running the easier workflow may not be feasible when cohorts exceed 200 samples, due to scatter job size limitations.
+[This notebook](../workflow/kf-joint-cohort-call-advanced-setup.ipynb) can be used to set up the split by chromosome jobs, in which N jobs are created to split X files per instance, depending on yor math.
```
```suggestion
[This notebook](../workflow/kf-joint-cohort-call-advanced-setup.ipynb) can be used to set up the split by chromosome jobs, in which N jobs are created to split X files per instance, depending on your math.
```

dmiller15 commented on docs/KF_NOTEBOOK_JG_WORKFLOW.md:
```diff
@@ -0,0 +1,5 @@
+# Kids First-Sentieon Joint Cohort Call Notebook
+Due to platform limitations, running the easier workflow may not be feasible when cohorts exceed 200 samples, due to scatter job size limitations.
+[This notebook](../workflow/kf-joint-cohort-call-advanced-setup.ipynb) can be used to set up the split by chromosome jobs, in which N jobs are created to split X files per instance, depending on yor math.
+Then with sensible tagging and sorting, the outputs from these jobs can be reliable pass to the Sentieon GVCFtyper to create the final product.
```
```suggestion
Then with sensible tagging and sorting, the outputs from these jobs can be reliably passed to the Sentieon GVCFtyper to create the final product.
```

dmiller15 commented on docs/KF_NOTEBOOK_JG_WORKFLOW.md:
```diff
@@ -0,0 +1,5 @@
+Due to platform limitations, running the easier workflow may not be feasible when cohorts exceed 200 samples, due to scatter job size limitations.
+[This notebook](../workflow/kf-joint-cohort-call-advanced-setup.ipynb) can be used to set up the split by chromosome jobs, in which N jobs are created to split X files per instance, depending on yor math.
+Then with sensible tagging and sorting, the outputs from these jobs can be reliable pass to the Sentieon GVCFtyper to create the final product.
+In the end, you'll want to delete all files with the `INTERMEDIATE` tag to avoid excessive storage costs on files that are likely cheaper to regenerate as needed than store indefinitely
```
```suggestion
In the end, you'll want to delete all files with the `INTERMEDIATE` tag to avoid excessive storage costs. These files can be quickly and cheaply regenerated, as needed.
```

dmiller15 commented on workflow/kf-joint-cohort-call-advanced-setup.ipynb:
You recommend chr1-22,X,Y,M here. The other documentation doesn't include M

dmiller15 reviewed (changes requested):
Overall, I think more work should be done to clean up the Python notebook. I'm not an expert on making those, but, if it's possible, I think we should have an opening block where users can put in the variable/unique-to-their-situation values then have it be pretty much hands off from there. Like our workflows: "just set these parameters, then we'll do the rest."

migbro committed:
Apply suggestions from code review

Co-authored-by: Dan Miller <dmiller15@users.noreply.github.com>

migbro committed:
:pencil: made doc fixes
:hamm:er refactored notebook with up-front variables and run instructions

migbro committed:
:pencil: fixed chr recs

migbro commented:
Ok, I refactored the notebook to have all needed variables set up front. Split jobs finished here:
https://cavatica.sbgenomics.com/u/d3b-bixu/kf-sentieon-join-cohort-genotyping-dev/tasks?search=Split%20GVCF%20by%20chr%20set
Joint jobs finished here: https://cavatica.sbgenomics.com/u/d3b-bixu/kf-sentieon-join-cohort-genotyping-dev/files/#q?search=CBTN-TEST_

dmiller15 reviewed (approved):
I slacked you a small refactor for one of the ipynb steps but otherwise looks good to me

migbro committed:
:hammer: added in Dan's refactored code block

wongjessica93 commented on docs/KF_COHORT_JG_SINGLE_WF.md:
```diff
@@ -0,0 +1,48 @@
+# [Kids First-Sentieon Joint Cohort Calling Workflow Beta](../workflow/kf-joint-cohort-call-by-chr-wf.cwl)
+This workflow is takes a set of gVCFs and an indexed reference FASTA to create a joint-called cohort VCF.
```
Typo - 
This workflow takes a set of gVCFs and an indexed reference FASTA to create a joint-called cohort VCF.

wongjessica93 commented on docs/KF_NOTEBOOK_JG_WORKFLOW.md:
```diff
@@ -0,0 +1,11 @@
+# Kids First-Sentieon Joint Cohort Call Notebook
+Due to platform limitations, running the Preferred workflow may not be feasible when cohorts exceed 2200 samples, due to scatter job size limitations.
```
Idk if this has an obvious answer, but what should users do if the cohort is > 2200? Should they break up the samples into smaller cohorts and do multiple runs? Or is the breaking up of chromosomes for the entire cohort the solution? It's unclear when reading this section

migbro replied:
Oh, well, it's to run this one up to a certain point, probably like 3500. Lemme double check, I put some hints elsewhere, but you get to a point where, for this platform, you'd have to either use smaller cohorts (but sometimes that's just not feasible as perhaps all samples have to be run together and can't be further sub-categorized), or focus on certain regions, which this workflow does not do. That might be a future update

migbro committed:
:pencil: added PR request for clarification and typo fixes
